log4j-spring-cloud-config/log4j-spring-cloud-config-client/src/main/java/org/apache/logging/log4j/spring/cloud/config/client/Log4j2CloudConfigLoggingSystem.java

```diff
@@ -0,0 +1,127 @@
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache license, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the license for the specific language governing permissions and
+ * limitations under the license.
+ */
+package org.apache.logging.log4j.spring.cloud.config.client;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.URLConnection;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Properties;
+import javax.net.ssl.HttpsURLConnection;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.core.LoggerContext;
+import org.apache.logging.log4j.core.config.ConfigurationFactory;
+import org.apache.logging.log4j.core.config.ConfigurationSource;
+import org.apache.logging.log4j.core.net.ssl.LaxHostnameVerifier;
+import org.apache.logging.log4j.core.net.ssl.SslConfiguration;
+import org.apache.logging.log4j.core.net.ssl.SslConfigurationFactory;
+import org.apache.logging.log4j.core.util.FileUtils;
+import org.apache.logging.log4j.status.StatusLogger;
+import org.apache.logging.log4j.util.PropertiesUtil;
+import org.springframework.boot.logging.LogFile;
+import org.springframework.boot.logging.LoggingInitializationContext;
+import org.springframework.boot.logging.log4j2.Log4J2LoggingSystem;
+import org.springframework.core.io.ClassPathResource;
+import org.springframework.util.Assert;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.ResourceUtils;
+
+/**
+ *
+ */
+public class Log4j2CloudConfigLoggingSystem extends Log4J2LoggingSystem {
+    private static final String FILE_PROTOCOL = "file";
+    private static final String HTTPS = "https";
+    private Logger LOGGER = StatusLogger.getLogger();
+
+    public Log4j2CloudConfigLoggingSystem(ClassLoader loader) {
+        super(loader);
+    }
+
+    @Override
+    protected String[] getStandardConfigLocations() {
+        String[] locations = super.getStandardConfigLocations();
+        PropertiesUtil props = new PropertiesUtil(new Properties());
+        String location = props.getStringProperty(ConfigurationFactory.CONFIGURATION_FILE_PROPERTY);
+        if (location != null) {
+            List<String> list = Arrays.asList(super.getStandardConfigLocations());
+            list.add(location);
+            locations = list.toArray(new String[list.size()]);
```

Comment:
From the latest benchmarks I remember reading about, it's generally faster to just pass 0 here instead of the real size since `toArray()` has to check `list.size()` first anyways.

Reply:
I had always thought that if you pass in an array that toArray would just use it.

Reply:
The key here is that even if you pass it an array, toArray has to check the length of the passed array against the length of the array being copied. toArray will resize the array if needed, so there's less if-checks executed in the array of length 0 execution path. Basically, toArray() is programmed to construct the new array regardless, so save yourself the bytecode.

Reply:
You must be thinking of a different method. The javadoc for this says "Returns a fixed-size list backed by the specified array. (Changes to the returned list "write through" to the array.)". Looking at the implementation that is exactly what it does. Passing an array of size 0 would require me to manually populate the list from the array, which is the whole point of this method.

Reply:
Oh I think you're right. Found more info [here](https://shipilev.net/blog/2016/arrays-wisdom-ancients/) and it doesn't seem to be an issue in this scenario (or in general anymore; this old issue seems to have been fixed). The key difference here is that calling `size()` on an array list is `O(1)`, so it's not as big an issue. If you were calling, say, `Collection::toArray` on a collection that doesn't cache its size (random glance at some standard collection classes indicates they normally do cache that), or worse, if the collection changed between checking the size and calling toArray, then that causes the performance latency.